misc/multistream-select/src/protocol.rs

```diff
@@ -0,0 +1,486 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Multistream-select protocol messages an I/O operations for
+//! constructing protocol negotiation flows.
+//!
+//! A protocol negotiation flow is constructed by using the
+//! `Stream` and `Sink` implementations of `MessageIO` and
+//! `MessageReader`.
+
+use bytes::{Bytes, BytesMut, BufMut};
+use crate::length_delimited::{LengthDelimited, LengthDelimitedReader};
+use futures::{prelude::*, try_ready};
+use log::trace;
+use std::{io, fmt, error::Error, convert::TryFrom};
+use tokio_io::{AsyncRead, AsyncWrite};
+use unsigned_varint as uvi;
+
+/// The maximum number of supported protocols that can be processed.
+const MAX_PROTOCOLS: usize = 1000;
+
+/// The maximum length (in bytes) of a protocol name.
+///
+/// This limit is necessary in order to be able to unambiguously parse
+/// response messages without knowledge of the corresponding request.
+/// 140 comes about from 3 * 47 = 141, where 47 is the ascii/utf8
```

Comment:
I don't really understand what the ASCII encoding of `/` has to do with the length.

Reply:
The issue was that if a "protocols list response" happens to contain 47 protocols, then its encoding starts with a uvi-length byte that resembles `/` and hence decoding decodes the entire thing as a single protocol message upon seeing the leading `/`. With the length restriction decoding is unambiguous because a "protocols list response" with 47 protocols is guaranteed to exceed that length and hence cannot be misinterpreted as a single protocol.

This is certainly not pretty but simply due to the fact that we have Sink/Stream abstractions over messages, so that decoding a message is unaware of the request in whose context decoding happens, i.e. there is no "expected" message when decoding. The protocol messages of multistream-select are unfortunately not designed for unambiguous decoding. We may want to go with a different approach of context-aware decoding in a future implementation, though I'm not sure if it can achieve the "elegance" of the Sink/Stream abstraction.

Let me know if my wording in the code comments can be made clearer.

Reply:
> We may want to go with a different approach of context-aware decoding in a future implementation, though I'm not sure if it can achieve the "elegance" of the Sink/Stream abstraction.

I think an approach using async/await can be extremely straight-forward, but for now it's ok.